### PR TITLE
Implement Tree list and new search UI

### PR DESCRIPTION
**Pull request by studioswong**
## Description

related PRs: [KED-2563](https://jira.quantumblack.com/browse/KED-2563), [KED-2564](https://jira.quantumblack.com/browse/KED-2564) and [KED-2564](https://jira.quantumblack.com/browse/KED-2659)

This is the main work in building a new tree-nav style list for navigating the kedro-viz sidebar arranged by modular pipelines, as well as the new search UI. ( see below). 

All elements ( tasks, dataset and parameters) are now arranged according to its reference in the nested structure of the modular pipelines. Each modular pipeline includes its children modular pipelines, along with their nested nodes arranged in the order of task nodes, followed by data and parameter nodes. 

Upon searching the list, the returned results would contain all nested nodes in an expanded nested format under their parent modular pipeline. 

tree nav demo 3

search-ui-small

note: I am opening this as a draft PR until I fix the failing tests for the JS import library tests

## Development notes

I have made the following changes:
1. set up a new component for the nested tree nav structure (`node-list-tree.js`)
2. set up another component to represent each item under the tree nav structure
3. set up a series of selectors for the new nested modular pipeline structure
4. updates to node-list for the split panel
5. refactoring of all node-list tests to cater for the new tree list ( please note that the related tests for the search UI and filter panel controls have been skipped and will be updated in the next PR for the search UI feature)
6. deleted the 'modular pipeline' field on the filter panel given modular pipelines are now exposed on the tree-list.
7. related style changes to cater for the tree-list
8. tests relating to all selectors under `node-list-items.test.js` for all newly added selectors. I noticed that there are some missing tests for selectors that have been set up previously, and I have also added some tests for them. 

## QA notes

With this feature my main focus is on setting up this new tree list and ensuring the search results are returned correctly. In terms of the UI, there are some uncertainties about the execution details for the new search UI, which we can further fine-tune on the design in future tasks. 

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

**Comment by studioswong** [comment]
Looks like some python tests are failing on the CI - @limdauto really appreciate your help to take a look :D Thank you so much!!

**Comment by limdauto** [comment]
@studioswong the Python failure came from the fact that it can't build a production bundle for the JS (the Python server test requires the bundle to be there since it tests API response with an index.html)

The failures are


```
Treating warnings as errors because process.env.CI = true.
Most CI servers set it automatically.

Failed to compile.

src/components/node-list/node-list-items.js
  Line 644:7:  'level' is assigned a value but never used  no-unused-vars

src/components/node-list/node-list-tree-search.js
  Line 85:6:  React Hook useEffect has missing dependencies: 'modularPipelineIds', 'modularPipelines', 'nodeModularPipelines', 'nodeSelected', 'nodeTypeIDs', and 'nodes'. Either include them or remove the dependency array  react-hooks/exhaustive-deps

src/selectors/modular-pipelines.js
  Line 3:10:  'arrayToObject' is defined but never used  no-unused-vars

```

If you follow the failure, you will see this in CI as well.

**Comment by studioswong** [comment]
> @studioswong the Python failure came from the fact that it can't build a production bundle for the JS (the Python server test requires the bundle to be there since it tests API response with an index.html)
> 
> The failures are
> 
> ```
> Treating warnings as errors because process.env.CI = true.
> Most CI servers set it automatically.
> 
> Failed to compile.
> 
> src/components/node-list/node-list-items.js
>   Line 644:7:  'level' is assigned a value but never used  no-unused-vars
> 
> src/components/node-list/node-list-tree-search.js
>   Line 85:6:  React Hook useEffect has missing dependencies: 'modularPipelineIds', 'modularPipelines', 'nodeModularPipelines', 'nodeSelected', 'nodeTypeIDs', and 'nodes'. Either include them or remove the dependency array  react-hooks/exhaustive-deps
> 
> src/selectors/modular-pipelines.js
>   Line 3:10:  'arrayToObject' is defined but never used  no-unused-vars
> ```
> 
> If you follow the failure, you will see this in CI as well.

indeed - thanks for highlighting that! my bad for not diving deeper in the CI. I'll tackle those errors.

**Review by limdauto — CHANGES_REQUESTED** [review]
Hey @studioswong thank you very much for the PR. It seems like it's working as expected when I test it out. Leaving my first batch of comments here.

File: src/components/node-list/node-list-items.js [hunk 1]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+    });
+
+    return modularPipelineNodes;
+  }
+);
+
+/**
+ * returns an array of modular pipelines arranged in a nested structure with corresponding nodes and names
+ */
+export const getNestedModularPipelines = createSelector(
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js:582]: yes indeed - I'll be adding more tests!

File: src/components/node-list/node-list-tree.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    nodes,
+    modularPipelines,
+    nodeSelected,
+    searchValue,
+    modularPipelineIds,
+    nodeModularPipelines,
+    nodeTypeIDs,
+  });
+
+  const searchTreeCount = getTreeSearchValueCount({ treeData, searchValue });
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-tree.js:72]: yes this is a selector I wrote for keeping count for the search results that will be shown on collapse the tree - I have kept this in for the potential next iteration, but since we are not including in this release I will take this out.

File: src/components/node-list/node-list-tree-search.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
+import React, { useState, useEffect } from 'react';
+import { connect } from 'react-redux';
+
+import { makeStyles, withStyles } from '@material-ui/core/styles';
+import TreeView from '@material-ui/lab/TreeView';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
+import ChevronRightIcon from '@material-ui/icons/ChevronRight';
+
```
Comment (rashidakanchwala) [inline src/components/node-list/node-list-tree-search.js:8]: @studioswong  -- I would like to understand a bit more on why do we use material-ui icons and not the svg arrows from Gabriel's design like we generally do for the rest of the icons in kedro?

File: src/components/node-list/node-list-tree.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  },
+  onToggleTypeDisabled: (typeID, disabled) => {
+    dispatch(toggleTypeDisabled(typeID, disabled));
+  },
+  onToggleNodeSelected: (nodeID) => {
+    dispatch(loadNodeData(nodeID));
+  },
+});
+
+export default connect(mapStateToProps, mapDispatchToProps)(TreeListProvider);
```
Comment (rashidakanchwala) [inline src/components/node-list/node-list-tree.js:179]: I am a little confused there are 2 TreeListProviders -- one in node-list-tree.js and one in node-list-tree-search.js. Is there code that needs to be removed?

**Comment by rashidakanchwala** [comment]
Hi, I have done the first review, didn't get a chance to go through all the tests but will have a look at them later today. It works smoothly at my end, so great work!!!! 
I think the two main things I noticed are that some of code is repetitive so if there's a way we can abstract it that would be nice. And the npm run test:coverage is not up to the mark but I think it's because there are tests still being written.

File: src/components/node-list/node-list.test.js [hunk 5]
```diff
@@ -21,7 +22,8 @@ describe('NodeList', () => {
     expect(nodeList.length).toBeGreaterThan(0);
   });
 
-  describe('search filter', () => {
+  // these tests are skipped for now as they will be tackled in the next PR with updated search features for the tree list
+  describe.skip('search filter', () => {
```
Comment (studioswong) [inline src/components/node-list/node-list.test.js:26]: sorry folks, I will be adding this test back in in my latest commit

File: src/components/node-list/node-list.test.js [hunk 6]
```diff
@@ -86,7 +88,9 @@ describe('NodeList', () => {
     });
   });
 
-  describe('visibility checkboxes on element items', () => {
+  // these tests refers to search and linked behaviour on the filter panel, which will be updated once the filter panel is merged
+  // into this PR
+  describe.skip('visibility checkboxes on element items', () => {
```
Comment (studioswong) [inline src/components/node-list/node-list.test.js:93]: same as above

File: src/components/node-list/node-list-tree.js (hunk 4 above)
Comment (studioswong) [reply src/components/node-list/node-list-tree.js:204]: I'm glad you raised this Rashida - the main reason for setting a new component for search is given the need to remount the `treeList` component to enable the search UI with the expanded tree ( otherwise it would generate a react error in regards to changing a controlled component to an uncontrolled element due to the unique setup of material-ui), and indeed it does causes repetitive code. 

I had separated the search-ui tree out as a separate component to provide a clear seperation of the search-ui logic away from the normal tree-ui (instead of incorporating it in the existing tree-list provider) so that I could easily strip that away once I find the solution to fix the issue with the react warning I mentioned above, which is a temporary solution hence the repetitive code. I should have added a note to explain that intention.

That said, i agree with what you mentioned above, and had refactored and incorporated the search logic into the original tree-list provider.

File: src/components/node-list/node-list-items.js [hunk 7]
```diff
@@ -148,15 +152,15 @@ export const getFilteredModularPipelineItems = createSelector(
...
+        visibleIcon: VisibleIcon,
+        invisibleIcon: InvisibleIcon,
         active: false,
         selected: false,
         faded: false,
         visible: true,
         disabled: false,
-        unset: !modularPipeline.enabled,
-        checked: modularPipeline.enabled,
+        unset: false,
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js:162]: thanks for raising this - this change is mainly due to the change of the modular pipeline item now being a 'element' type item rather than 'filter' type similar to tag, where the unset field is basically only used in a scenario similar to setting an individual tag, which does not exist anymore with this iteration. 

( i.e `modularPipeline.enabled` for each modular pipeline is not controlled by setting in the same fashion as setting a filter on or off) hence I am setting this false by default to avoid confusion, in a similar fashion to how the node items are constructed ( refer to line 225 to see how we have set the `unset` field for node items as default.)

File: src/components/node-list/node-list-tree-search.js (hunk 3 above)
Comment (studioswong) [reply src/components/node-list/node-list-tree-search.js]: the arrow used here from Gabriel's design is a generic  'expand more' icon ( the icon used in the tree is different from the other expand more icon used on the sidebar) - given that that icon is only used here, and since material-ui provides that out of the box, I opted for the simplest solution to use their icon. This also avoids the need to particularly maintain more code for a one-off icon just for the tree.

**Comment by rashidakanchwala** [comment]
@studioswong  -- Great work! Seem's like you have addressed my comments above. I understand the issues you were facing with react components and hence the need for structuring the code the way you did. Looks fine to me and thanks for adding comments in the code for some of the stuff. It's make it easier to understand the reasons for implementation.

**Review by rashidakanchwala — APPROVED** [review]

File: src/components/node-list/node-list-items.js [hunk 8]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+    filteredModularPipelines,
+    modularPipelines,
+    nodeTypeIDs
+  ) => {
+    const filteredNodeMoudularPipelines = [];
+    const filteredMP = filteredModularPipelines.modularPipeline;
+
+    const nodeItems = Object.assign({}, filteredNodeItems);
+
+    nodeTypeIDs.forEach((key) => {
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js:375]: Thanks Lim, I just saw this comment from the `files changed` page on github while this didn't show up on the main page for this PR. 

I'm glad that you raised this - `filteredNodeItems` is not an array but rather an object that is arranged in groups determined by `nodeTypeIds` ( which in our case, is `task`, `parameters`, and `dataset`). Hence this is why we need to 

On top of this, I a

File: src/components/node-list/node-list-items.js [hunk 9]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+    (state) => state.nodeTypeIDs,
+  ],
+  (
+    filteredNodeItems,
+    filteredModularPipelines,
+    modularPipelines,
+    nodeTypeIDs
+  ) => {
+    const filteredNodeMoudularPipelines = [];
+    const filteredMP = filteredModularPipelines.modularPipeline;
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: yes it is - this is an unfortunate situation due to the previous setup where `filteredModularPipelines` returns the pipelines array under an additional `modularPipeline` key - here I am trying to introduce non-breaking code given that Liam's filter work would also rely on this selector, hence I kept it in this rather than modifying the previous selector.

File: src/components/node-list/node-list.test.js (hunk 6 above)
Comment (studioswong) [reply src/components/node-list/node-list.test.js]: the above test is now resolved in my latest commit with modification to match the latest tree-search UI

File: src/components/node-list/node-list-items.js [hunk 10]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+      });
+    });
+
+    return filteredNodeMoudularPipelines;
+  }
+);
+
+/**
+ * constructs a modular pipeline item for filtered modualar pipeline parents that does not exist in filtered modualr pipeline items
+ * @param {obj modularPipeline the modular pipeine that needs the construction of a modular pipeline item
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: This has been updated to reflect the suggested change

File: src/components/node-list/node-list-items.js [hunk 11]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+    (state) => state.modularPipelines,
+    (state) => state.nodeTypeIDs,
+  ],
+  (
+    filteredNodeItems,
+    filteredModularPipelines,
+    modularPipelines,
+    nodeTypeIDs
+  ) => {
+    const filteredNodeMoudularPipelines = [];
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: This has been updated to reflect the suggested change

File: src/components/node-list/node-list-items.js [hunk 12]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+
+    const nodeItems = Object.assign({}, filteredNodeItems);
+
+    nodeTypeIDs.forEach((key) => {
+      nodeItems[key]?.forEach((filteredNode) => {
+        filteredNode.modularPipelines.forEach((nodeModularPipeline) => {
+          if (!filteredMP.some((mpItem) => mpItem.id === nodeModularPipeline)) {
+          }
+
+          filteredNodeMoudularPipelines.push(
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: This has been updated to reflect the suggested change

File: src/components/node-list/node-list-items.js [hunk 13]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+        });
+      });
+    });
+
+    return filteredNodeMoudularPipelines;
+  }
+);
+
+/**
+ * constructs a modular pipeline item for filtered modualar pipeline parents that does not exist in filtered modualr pipeline items
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: This has been updated to reflect the suggested change

File: src/components/node-list/node-list-items.js [hunk 14]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+ * @param {obj modularPipeline the modular pipeine that needs the construction of a modular pipeline item
+ * @return {boolean} True if match
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: This has been updated to reflect the suggested change

File: src/components/node-list/node-list-items.js [hunk 15]
```diff
@@ -330,3 +350,333 @@ export const getFilteredItems = createSelector(
...
+    modularPipelineParent,
+    modularPipelines
+  ) => {
+    modularPipelineItems = modularPipelineItems.modularPipeline;
+
+    let finalModularPipelines = [];
+
+    // sort all 3 sets of modular pipelines according to the original order
+    modularPipelines?.forEach((mp) => {
+      // check filtered modualr pipelines
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js]: This has been updated to reflect the suggested change

**Review by limdauto — APPROVED** [review]
LGTM! Thank you so much for pushing this over the line.

**Review by bru5 — CHANGES_REQUESTED** [review]
As discussed looks like there are some design issues to resolve. I'll continue looking at the code meanwhile and get back.

File: src/components/node-list/node-list-items.js [hunk 16]
```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
...
+        lastIndex = childrenModularPipeline.id.indexOf('.', lastIndex) + 1;
+      }
+    });
+
+    return filteredMoudularPipelineParents;
+  }
+);
+
+/**
+ * returns the corresponding final set of modular pipelines
```
Comment (bru5) [inline src/components/node-list/node-list-items.js:475]: Where practical please try use JSDoc format (as used elsewhere) - I know this can be tricky and some places have it missing, but still worth trying to keep it up especially for important functions like this, as well as maintaining any existing docs when changing things.

File: src/components/node-list/node-list-items.js [hunk 17]
```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
...
+  ) => {
+    modularPipelineItems = modularPipelineItems.modularPipeline;
+
+    let finalModularPipelines = [];
+
+    // sort all 3 sets of modular pipelines according to the original order
+    modularPipelines?.forEach((modularPipeline) => {
+      // check filtered modular pipelines
+      if (
+        modularPipelineItems.some(
```
Comment (bru5) [inline src/components/node-list/node-list-items.js:499]: Similar to earlier, I think pull these conditions out of the `if` statements and give them a variable with a helpful name otherwise it's really hard to see what's a condition and what's assignment etc.

File: src/components/node-list/node-list-items.js [hunk 18]
```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
...
+              modularPipelineItem.id === modularPipeline.id
+          )[0]
+        );
+      } else if (
+        nodeModularPipelines.some(
+          (nodeModularPipeline) => nodeModularPipeline.id === modularPipeline.id
+        )
+      ) {
+        finalModularPipelines.push(
+          nodeModularPipelines.filter(
```
Comment (bru5) [inline src/components/node-list/node-list-items.js:515]: Worth considering using `.find(...)` rather than `.filter(...)[0]` for this sort of thing.

File: src/components/node-list/node-list-items.js [hunk 19]
```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
...
+  [
+    getFilteredNodeItems,
+    getFilteredTreeItems,
+    (state) => state.modularPipelineIds,
+    (state) => state.nodeTypeIDs,
+  ],
+  (filteredNodeItems, filteredTreeItems, modularPipelineIDs, nodeTypeIDs) => {
+    const modularPipelineNodes = arrayToObject(modularPipelineIDs, () => []);
+
+    const nodeItems = cloneDeep(filteredNodeItems);
```
Comment (bru5) [inline src/components/node-list/node-list-items.js:553]: Is there away to avoid `cloneDeep`? It's generally not a great idea for performance and can lead to data integrity bugs / confusion, so maybe there's another way?

File: src/components/node-list/node-list-tree.js [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+            searchValue,
+            modularPipelineIds,
+            nodeModularPipelines,
+            nodeTypeIDs,
+          })
+        : [];
+
+    let expandedModularPipelines = [];
+
+    searchValue !== '' &&
```
Comment (bru5) [inline src/components/node-list/node-list-tree.js:83]: How come there is a conditional here? Should this be inside an if statement instead? Same goes for the above lines in this `useEffect`?

File: src/components/node-list/node-list-tree.js [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+      onItemClick={onItemClick}
+      key={rowData.id}>
+      {renderModularPipelines(rowData)}
+
+      {/* render set of node elements in that modular pipeline */}
+      {renderChildNodes(rowData)}
+    </NodeListTreeItem>
+  );
+
+  return searching ? (
```
Comment (bru5) [inline src/components/node-list/node-list-tree.js:158]: These look roughly the same, can just one `StyledTreeView` be used here and then conditionals on the attributes instead?

Just thought worth noting here too that I'd say usually prefer to go for an `if` statement block first over using the ternary `?` where it's not a short and trivial assignment.

**Review by bru5 — CHANGES_REQUESTED** [review]

File: src/components/node-list/node-list-items.js [hunk 22]
```diff
@@ -175,6 +179,21 @@ const compareEnabledThenAlpha = (itemA, itemB) => {
...
+/**
+ * Compares items for sorting in each modularPipeline in the nested tree structure
+ * by enabled status (by tag) and then alphabeticaly (by type)
+ * @param {object} itemA First item to compare
+ * @param {object} itemB Second item to compare
+ * @return {number} Comparison result
+ */
+const compareEnabledThenType = (itemA, itemB) => {
+  const byEnabledTag = Number(itemA.disabled_tag) - Number(itemB.disabled_tag);
+  const nodeTypeIDs = ['task', 'data', 'parameters'];
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js:191]: There was no need to sort by type before as each type was displayed as a separate list on the sidebar in the old design hence there was no sorting done at all by type in our previous execution(see below). 

Without this sorting, the elements would not be arranged in this order within the modular pipeline.  ( you can try it by taking this function out.) This is done given the additional requirement raised by you and gabriel for sorting the elements by this type order in our previous discussions. 
image

File: src/components/node-list/node-list-items.js (hunk 19 above)
Comment (studioswong) [reply src/components/node-list/node-list-items.js:553]: There is a need to use deepClone here actually to avoid bugs down the line - the item referenced here for `filteredNodeItems` will be used in other selectors later on and hence requires immutability. I had previously used shallow clone methods such as `Object.assign` and there proved to be problems down the line with the other selectors. 

So in short, deep clone is neccessary here. The approach and preference to use `cloneDeep` from lodash in our codebase was also advocated by Richard in his previous PRs, hence I am using it in this instance.

File: src/components/node-list/node-list-tree.js (hunk 20 above)
Comment (studioswong) [reply src/components/node-list/node-list-tree.js:83]: This is to detect the cases for empty searchValue in the search bar - this is also demonstrated in our legacy tests throughout the code base in detecting the case of empty value in the search bar

File: src/components/node-list/node-list-tree.js (hunk 21 above)
Comment (studioswong) [reply src/components/node-list/node-list-tree.js:158]: haha yea in respond to this - please see my previous answer to Rashida's question. The main reason we need to specify a seperate `StyledTreeView` is the need to reload that in going between the normal tree-view and the search Tree, which is an expanded tree. ( note the need to specify the `expanded` prop for the searching condition). Passing a new value in between the searching and the non-search scenerio would cause a react error on the browser as we are forcing to switch between a controlled and uncontrolled react component.) Doing so would force the component to remount and allow stability in performance, as well as to elinate the react error.

File: src/components/node-list/node-list-items.js [hunk 23]
```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
...
+    getFilteredNodeItems,
+    (state) => state.modularPipelines,
+    (state) => state.nodeTypeIDs,
+  ],
+  (filteredNodeItems, modularPipelines, nodeTypeIDs) => {
+    const filteredNodeModularPipelines = [];
+
+    const nodeItems = cloneDeep(filteredNodeItems);
+
+    nodeTypeIDs.forEach((nodeTypeId) => {
```
[the remark this answers is not in the document]
Comment (studioswong) [reply src/components/node-list/node-list-items.js:368]: I'm not entirely sure what you meant by flatten this? Arranging it in this arrangement is actually more readable (as in response to your previous comments about readability, arranging it this way will show the clear hierarchy in how we are iterating the filtered items according to the order listed out from `nodeTypeIds`).

File: src/components/node-list/node-list-items.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (bru5) [reply src/components/node-list/node-list-items.js:191]: Ah yes, the ordering used to come out of the config then I think, thanks for clarifying.

File: src/components/node-list/node-list-tree.js (hunk 20 above)
Comment (bru5) [reply src/components/node-list/node-list-tree.js:83]: Yeah I get what you mean, I just meant it should probably be inside an `if (...) { ... } ` otherwise it's not easy to see that it's a conditional?

File: src/components/node-list/node-list-tree.js (hunk 21 above)
Comment (bru5) [reply src/components/node-list/node-list-tree.js:158]: Ah okay if it's a limitation of the framework, I guess worth adding that in a comment here?

File: src/components/node-list/node-list-items.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (bru5) [reply src/components/node-list/node-list-items.js:368]: Cool that's fine then, just checking thanks for looking.

File: src/components/node-list/node-list-items.js (hunk 19 above)
Comment (bru5) [reply src/components/node-list/node-list-items.js:553]: Okay cool got it, thanks for clarifying.

**Review by bru5 — APPROVED** [review]
Thanks for getting back on the questions, much appreciated in helping me understand this!
